mpi4jax/_src/collective_ops/sendrecv.py

```diff
@@ -234,10 +234,68 @@ def mpi_sendrecv_abstract_eval(
     )
 
 
+def mpi_sendrecv_value_and_jvp(
+    in_args, tan_args, source, dest, sendtag, recvtag, comm, status
+):
+    sendbuf, recvbuf, token = in_args
+    send_tan, recv_tan, token_tan = tan_args
+
+    val, token = mpi_sendrecv_p.bind(
+        sendbuf,
+        recvbuf,
+        token,
+        source=source,
+        dest=dest,
+        sendtag=sendtag,
+        recvtag=recvtag,
+        comm=comm,
+        status=status,
+    )
+
+    # throw away return token to work around jax#6285
+    jvp, token_jvp = mpi_sendrecv_p.bind(
+        send_tan,
+        recv_tan,
+        token,
+        source=source,
+        dest=dest,
+        sendtag=sendtag,
+        recvtag=recvtag,
+        comm=comm,
+        status=status,
+    )
+
+    return (val, token), (jvp, ad.Zero.from_value(token_jvp))
+
+
+def mpi_sendrecv_transpose_rule(
+    tan_args, *x_args, source, dest, sendtag, recvtag, comm, status
+):
+    _, _, token = x_args
+    out_tan, token_tan = tan_args
+
+    # swap the sender and receiver
+    res, token = mpi_sendrecv_p.bind(
+        out_tan,
+        out_tan,
+        token,
+        source=dest,
+        dest=source,
+        sendtag=sendtag,  # TODO: could maybe be smarter about send and receive tags...
+        recvtag=recvtag,  # TODO: could maybe be smarter about send and receive tags...
```

Comment:
I think the tags are fine. What's your issue with them?

Reply:
I'm unsure what tags are for, actually.
If they are to avoid deadlocks between messages sent/received or discriminating messages...?

Since this creates effectively another send_recv, maybe we should have send and recv tags for the tangents as inputs... (but it hsould be specified in the rule above, not here)

Reply:
I think tags are for situations like this:

```
if rank == 0:
    recv(..., source=1)
    recv(..., source=2)
elif rank == 1 or rank == 2:
    send(..., dest=0)
```

If the message from rank 2 arrives first, this would deadlock without tags. But with tags, it is able to use the right message.

Re-using the same tags as in the forward pass should be the correct thing to do.

Reply:
Ok, i'll delete the comment